common/raw_hashtable.h

```diff
@@ -0,0 +1,1206 @@
+    return &small_storage_;
+  }
+
+  mutable SmallStorage small_storage_;
+};
+
+// A template specialization for when no small size optimization buffer is
+// desired.
+template <typename InputBaseT>
+class TableImpl<InputBaseT, 0> : public InputBaseT {
+ protected:
+  using BaseT = InputBaseT;
+
+  TableImpl() : BaseT(0, nullptr) {}
+  TableImpl(const TableImpl& arg) : TableImpl() { this->CopyFrom(arg); }
+  template <ssize_t OtherMinSmallSize>
+  explicit TableImpl(const TableImpl<BaseT, OtherMinSmallSize>& arg)
+      : TableImpl() {
+    this->CopyFrom(arg);
+  }
+  TableImpl(TableImpl&& arg) noexcept : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+  template <ssize_t OtherMinSmallSize>
+  explicit TableImpl(TableImpl<BaseT, OtherMinSmallSize>&& arg) : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+
+  auto ResetImpl() -> void;
+};
+
+// Computes a seed that provides a small amount of entropy from ASLR where
+// possible with minimal cost. The priority is speed, and this computes the
+// entropy in a way that doesn't require loading from memory, merely accessing
+// entropy already available without accessing memory.
+inline auto ComputeSeed() -> uint64_t {
+  // A global variable whose address is used as a seed. This allows ASLR to
+  // introduce some variation in hashtable ordering when enabled via the code
+  // model for globals.
+  extern volatile std::byte global_addr_seed;
+
+  return reinterpret_cast<uint64_t>(&global_addr_seed);
+}
+
+inline auto ComputeProbeMaskFromSize(ssize_t size) -> size_t {
+  CARBON_DCHECK(llvm::isPowerOf2_64(size))
+      << "Size must be a power of two for a hashed buffer!";
+  // The probe mask needs to mask down to keep the index within
+  // `size`. Since `size` is a power of two, this is equivalent to
+  // `size - 1`. We also mask off the low bits while here to match the size of
+  // the groups of entries.
+  return (size - 1) & ~GroupMask;
+}
+
+// This class handles building a sequence of probe indices from a given
+// starting point, including both the quadratic growth and masking the index
+// to stay within the bucket array size. The starting point doesn't need to be
+// clamped to the size ahead of time (or even by positive), we will do it
+// internally.
+//
+// We compute the quadratic probe index incrementally, but we can also compute
+// it mathematically and will check that the incremental result matches our
+// mathematical expectation. We use the quadratic probing formula of:
+//
+//   p(x,s) = (x + (s + s^2) / 2) mod (Size / GroupSize)
+//
+// This particular quadratic sequence will visit every value modulo the
+// provided size divided by the group size.
+//
+// However, we compute it scaled to the group size constant G and have it visit
+// each G multiple modulo the size using the scaled formula:
+//
+//   p(x,s) = (x + (s + (s^2 * G) / G^2) / 2) mod Size
+class ProbeSequence {
+  ssize_t Step = 0;
+  size_t Mask;
+  ssize_t i;
```

Comment:
This seems to be `p` from the comment?
```suggestion
  ssize_t p;
```

Reply:
Yep.